rfd/0047-drop-vendor.md

```diff
@@ -0,0 +1,156 @@
+that is [run by Google](https://proxy.golang.org) until we determine that
+running our own proxy is warranted.
+
+### Developer Experience
+
+Committing the vendor directory has led to a number of minor issues, and
+dropping it would result in an improved developer experience.
+
+First, comitting these changes results in repo bloat. Since go.mod and go.sum
+already contain everything needed for a reproducible build, the teleport repo
+ends up being larger than necessary. This causes clone operations to take longer,
+increases the time taken for CI builds, etc.
+
+Second, changes that bump dependencies are often large and difficult to review.
+This is especially true if a change includes updates to dependencies and
+teleport source, as reviewers may have to parse through large dependency diffs
+in order to find the core teleport changes.
+
+Additionally, our approach for the `api` module and vendoring requires a
+particular workflow. We leverage a module replace directive, which allows us to
+provide a versioned API module to external developers, while always building
+against latest for official builds. Unfortunately, replace directives don't play
+nice with vendoring, so we use an
+[error-prone symbolic link](https://github.com/gravitational/teleport/blob/30effc1f08b6a699772ff22f79ebe756fe1a1e34/Makefile#L942-L952)
+which is difficult to maintain in a cross-platform way, and has broken
+[gopls integration](https://github.com/gravitational/teleport/blob/30effc1f08b6a699772ff22f79ebe756fe1a1e34/Makefile#L942-L952)
+a common tool used in Go development environments.
+
+Lastly, there is no guarantee that the code committed to vendor actually
+reflects the contents of go.mod. The onus is on the developer to remember to run
+`make update-vendor` and commit the results after making changes to
+dependencies. This has created several cases of confusing build results amongst
+the dev team in recent history.
+
+In summary, dropping vendor would make go.mod/go.sum the single source of truth
+for our dependencies, reduce the size of the teleport repo, make code changes
+easier to review, and restore functionality with the latest version of `gopls`.
+
+### Impact to CI
+
+If the vendor directory is dropped, each CI build will need to pull down
+dependencies on first use. Should this become an issue, we can cache
+the Go module cache in [Google Cloud Storage](https://cloud.google.com/build/docs/speeding-up-builds)
+to speed up builds.
+
+Also note that while Google Cloud Build performs a shallow checkout by default,
+we are currently using `--unshallow` to fetch the entire history, so we will not
+realize any benefits in terms of clone time. As of the time of this writing,
+a full clone of the repository consumes 30-40 seconds of a 10+ minute build,
+so there's unlikely to be a need to change this behavior.
+
+### Security
+
+The removal of the vendor directory does eliminate one workflow that is nice
```

Comment:
Is it worth giving some concrete examples of what happens if someone should attempt a supply chain attack, and how the modules system will stop it? 

For example, given the scanario:

1. Daphne publishes the module `mystery-machine.com/scooby/v2`
2. Fred uses `scooby/v2` in Teleport
3. Evil Mr Winkles (running a land-grab scam of some sort) somehow gets control of the `mystery-machine` repo and  re-publishes a compromised `scooby/v2`
4. Fred rebuilds Teleport

It might be worth explaining at what point (and by what mechanism) does the modules system stop the compromised `scooby/v2` getting into the Teleport binary

Or even just a link to one would be useful.

Reply:
Sure, it could be useful, I'll try to add some context here.

Worth pointing out though - nothing really changes as to the behavior we have today. The same module system is running when we update a dependency and `make update-vendor` - we're just no longer committing the result.